File: src/main/java/komu/blunt/eval/RootBindings.java
package komu.blunt.eval;

import komu.blunt.analyzer.StaticEnvironment;
import komu.blunt.analyzer.VariableReference;
import komu.blunt.objects.Symbol;
import komu.blunt.types.DataTypeDefinitions;
import komu.blunt.types.Scheme;
import komu.blunt.types.checker.Assumptions;

import static komu.blunt.objects.Symbol.symbol;

public final class RootBindings {
    public final StaticEnvironment staticEnvironment = new StaticEnvironment();
    final RootEnvironment runtimeEnvironment = new RootEnvironment();
    private final Assumptions.Builder assumptions = Assumptions.builder();
    public final DataTypeDefinitions dataTypes = new DataTypeDefinitions();

    public void bind(String name, Scheme type, Object value) {
        bind(symbol(name), type, value);
    }

    public void bind(Symbol name, Scheme scheme, Object value) {
        VariableReference ref = staticEnvironment.define(name);
        defineVariableType(name, scheme);
        runtimeEnvironment.define(ref, value);
    }

    public void defineVariableType(Symbol name, Scheme scheme) {
        assumptions.add(name, scheme);
    }

    public Assumptions createAssumptions() {
        return assumptions.build();
    }
}

File: src/main/kotlin/komu/blunt/eval/RootBindings.kt
package komu.blunt.eval

import komu.blunt.analyzer.StaticEnvironment
import komu.blunt.objects.Symbol
import komu.blunt.parser.TypeParser
import komu.blunt.types.DataTypeDefinitions
import komu.blunt.types.Scheme
import komu.blunt.types.checker.Assumptions

class RootBindings {
    val staticEnvironment = StaticEnvironment()
    val runtimeEnvironment = RootEnvironment()
    private val assumptions = Assumptions.builder()
    val dataTypes = DataTypeDefinitions()

    fun bind(name: String, scheme: Scheme, value: Any?) {
        bind(Symbol(name), scheme, value)
    }

    fun bind(name: Symbol, scheme: Scheme, value: Any?) {
        val ref = staticEnvironment.define(name)
        defineVariableType(name, scheme)
        runtimeEnvironment.define(ref, value)
    }

    fun bind(name: String, scheme: String, value: Any?) {
        bind(name, TypeParser.parseScheme(scheme), value)
    }

    fun bindFunction(name: String, scheme: String, value: (Any?) -> Any?) {
        bind(name, TypeParser.parseScheme(scheme), value)
    }

    fun defineVariableType(name: Symbol, scheme: Scheme) {
        assumptions.add(name, scheme)
    }

    fun createAssumptions(): Assumptions =
        assumptions.build()
}
